internal/typeinfo/valuelocator_test.go

```diff
@@ -127,162 +127,266 @@ func (s *typeInfoSuite) TestLocateScanTargetError(c *C) {
+		argInfo, err := GenerateArgInfo([]any{t.typeSample})
+		c.Assert(err, IsNil)
 
-	for _, test := range tests {
-		valOfSlice := reflect.ValueOf(test.slice)
 		typeToValue := map[reflect.Type]reflect.Value{
-			reflect.TypeOf(test.slice): valOfSlice,
+			reflect.TypeOf(t.arg): reflect.ValueOf(t.arg),
 		}
-
-		input, err := argInfo.InputSlice(valOfSlice.Type().Name())
+		input, err := t.input(argInfo)
 		c.Assert(err, IsNil)
 
-		vals, omit, err := input.LocateParams(typeToValue)
+		params, err := input.LocateParams(typeToValue)
 		c.Assert(err, IsNil)
-		c.Assert(omit, Equals, false)
-		c.Assert(vals, HasLen, len(test.expectedValues))
-
-		for i := 0; i < len(test.expectedValues); i++ {
-			c.Assert(vals[i].Interface(), Equals, test.expectedValues[i])
+		c.Check(params.Omit, Equals, t.expectedOmit)
+		c.Check(params.Bulk, Equals, t.expectedBulk)
+		c.Assert(params.Vals, HasLen, len(t.expectedVals))
+		for i, v := range t.expectedVals {
+			c.Check(params.Vals[i], Equals, v)
 		}
 	}
 }
 
-func (*typeInfoSuite) TestLocateParamsSliceError(c *C) {
-	type S []any
-	type T []int
+func (s *typeInfoSuite) TestLocateParamsError(c *C) {
+	tests := []struct {
+		summary    string
+		typeSample any
+		arg        any
+		input      func(ArgInfo) (Input, error)
+		err        string
+	}{{
+		summary:    "invalid map key",
+		typeSample: M{},
+		arg:        M{"foo": "bar"},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "baz")
+		},
+		err: `map "M" does not contain key "baz"`,
+	}, {
+		summary:    "missing map type",
+		typeSample: M{},
+		arg:        N{"foo": "bar"},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "baz")
+		},
+		err: `parameter with type "M" missing (have "N")`,
+	}, {
+		summary:    "missing struct type",
+		typeSample: TS{},
+		arg:        TT{},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("TS", "foo")
+		},
+		err: `parameter with type "TS" missing (have "TT")`,
+	}, {
+		summary:    "slice not found",
+		typeSample: Sint{},
+		arg:        S{},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputSlice("Sint")
+		},
+		err: `parameter with type "Sint" missing (have "S")`,
+	}, {
+		summary:    "map bulk insert invalid key",
+		typeSample: M{},
+		arg:        []M{{"foo": "bar"}},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "baz")
+		},
+		err: `map "M" does not contain key "baz"`,
+	}, {
+		summary:    "map bulk insert nil map in slice",
+		typeSample: M{},
+		arg:        []M{{"foo": "bar"}, nil},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "foo")
+		},
+		err: `got nil map in slice of "M" at position 1`,
+	}, {
+		summary:    "map bulk insert nil pointer to map in slice",
+		typeSample: M{},
+		arg:        []*M{{"foo": "bar"}, nil},
+		input: func(ai ArgInfo) (Input, error) {
+			return ai.InputMember("M", "foo")
+		},
+		err: `got nil pointer in slice of "M" at position 1`,
```

Comment:
Should we say `index` instead of position? Index to me starts at 0 clearly.

Reply:
Changed